File: Macro/src/com/sioux/GameEngine/Planet.java
package com.sioux.GameEngine;


/**
 * Created by Michael on 3/15/2017.
 */
public class Planet {
    private int _id;
    private String _name;
    private int _orbitDistance;
    private int _orbitRotation;
    private int _ownedBy;

    public Planet(int id, String name, int orbitDistance, int orbitRotation, int ownedBy) {
        this._id = id;
        this._name = name;
        this._orbitDistance = orbitDistance;
        this._orbitRotation = orbitRotation;
        this._ownedBy = ownedBy;
    }

    public int get_id() { return this._id; }
    public void set_id(int id) { this._id = id; }

    public String get_name() { return this._name; }
    public void set_name(String name) { this._name = name; }

    public void set_orbitDistance(int _orbitDistance) { this._orbitDistance = _orbitDistance; }
    public int get_orbitDistance() { return this._orbitDistance; }

    public void set_orbitRotation(int _orbitRotation) { this._orbitRotation = _orbitRotation; }
    public int get_orbitRotation() { return this._orbitRotation; }

    public void set_ownedBy(int _ownedBy) { this._ownedBy = _ownedBy; }
    public int get_ownedBy() { return this._ownedBy; }
}

File: MacroEngine/Planet.cpp
#include "Planet.h"
#include "Player.h"

#include <QJsonObject>
#include <QColor>
#define _USE_MATH_DEFINES
#include <math.h>


Planet::Planet(QString name, int orbitDistance, double orbitRotation, double orbitSpeed, QObject *parent)
    : GameObject(parent)
    , m_name(name)
    , m_orbitDistance(orbitDistance)
    , m_orbitRotation(orbitRotation)
    , m_orbitSpeed(orbitSpeed)
    , m_ownedBy(-1)
{
}

void Planet::writeState(QJsonObject& gameState) const
{
    gameState["id"] = m_id;
    gameState["name"] = m_name;
    gameState["orbitDistance"] = m_orbitDistance;
    gameState["orbitRotation"] = m_orbitRotation;
    gameState["ownedBy"] = m_ownedBy;
    gameState["color"] = m_color.name(QColor::HexArgb);
}

void Planet::applyTick(double durationInSeconds)
{
    m_orbitRotation += 360.0*durationInSeconds / m_orbitSpeed;
}

int Planet::getOwnedBy() const
{
    return m_ownedBy;
}

void Planet::takeOverBy(Player* player)
{
    if (player)
    {
        m_ownedBy = player->getId();
        m_color = player->getColor();
    }
}

double Planet::getOrbitRotation() const
{
    return m_orbitRotation;
}

int Planet::getOrbitDistance() const
{
    return m_orbitDistance;
}

QPointF Planet::getCartesianCoordinates() const
{
    return QPointF(m_orbitDistance * cos( M_PI * (m_orbitRotation / 180.0)), -m_orbitDistance * sin(M_PI * (m_orbitRotation/ 180.0)));
}
